Implement PR #1682 - TPA Event

DrewAMitchell opened the pull request:
https://github.com/EssentialsX/Essentials/issues/1682
First pass at this solution, awaiting feedback. I've implemented the event, but was unsure what to do in the event that it was cancelled, so I made it to where neither party of the TPA request will receive any messages.
Another potential solution is only have the requester receive a message, similar to if they're being ignored.
Also was unsure of what to do for TPAAll, so give me feedback on that as well.

drtshock commented on Essentials/src/com/earth2me/essentials/commands/Commandtpa.java:
```diff
@@ -35,8 +38,12 @@ public void run(Server server, User user, String commandLabel, String[] args) th
+        TPARequestEvent tpaEvent = new TPARequestEvent(user.getSource(), player, false);
         if (!player.isIgnoredPlayer(user)) {
+        	ess.getServer().getPluginManager().callEvent(tpaEvent);
+            if(tpaEvent.isCancelled()) {
```
format

drtshock replied:
Also, why are we constructing the event without knowing if we're going to use it for sure or not?

DrewAMitchell replied:
Apologies, aforementioned issues will be fixed in the next commit.

drtshock commented on Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java:
```diff
@@ -35,7 +38,12 @@ public void run(final Server server, final User user, final String commandLabel,
         }
+        TPARequestEvent tpaEvent = new TPARequestEvent(user.getSource(), player, true);
         if (!player.isIgnoredPlayer(user)) {
+        	ess.getServer().getPluginManager().callEvent(tpaEvent);
```
format again

DrewAMitchell replied:
I hope this isn't a line spacing problem - I'll fix it in a future commit if it is, just review this next commit and let me know if it's still there.

drtshock commented on Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java:
```diff
@@ -42,6 +45,11 @@ private void teleportAAllPlayers(final Server server, final CommandSource sender
             }
             try {
+            	TPARequestEvent tpaEvent = new TPARequestEvent(sender, player, true);
+                ess.getServer().getPluginManager().callEvent(tpaEvent);
```
line spacing..

DrewAMitchell replied:
It was caused by tab indentation, will fix in the next commit.

drtshock reviewed (changes requested):
On top of what's in my comments, what's the reason for not extending one of our current events? https://github.com/EssentialsX/Essentials/blob/2.x/Essentials/src/net/ess3/api/events/FlyStatusChangeEvent.java

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java:
```diff
@@ -42,6 +45,11 @@ private void teleportAAllPlayers(final Server server, final CommandSource sender
+            	TPARequestEvent tpaEvent = new TPARequestEvent(sender, player, true);
+                ess.getServer().getPluginManager().callEvent(tpaEvent);
+                if(tpaEvent.isCancelled()) {
+                	return; // Potentially add a message letting requester know their request was cancelled?
```
I think that a plugin should let the user know if it cancelled the plugin and why, though EssentialsX *could* output a generic "Your teleport request was cancelled." message - I'd need to see how other EssentialsX events do this.

DrewAMitchell replied:
Will wait for further advice on this, willing to go either way.

DrewAMitchell replied:
Commit will include a teleport request cancelled message, unless you've got another recommendation.

mdcfe commented on Essentials/src/net/ess3/api/events/TPARequestEvent.java:
```diff
@@ -0,0 +1,55 @@
+
+
+public class TPARequestEvent extends Event implements Cancellable {
+	private static final HandlerList handlers = new HandlerList();
```
EssentialsX does not use tab indents.

DrewAMitchell replied:
Dually noted, will be converted in next commit.

mdcfe commented on Essentials/src/net/ess3/api/events/TPARequestEvent.java:
```diff
@@ -0,0 +1,55 @@
+		return requested;
+	}
+	
+	public boolean isRequestToTeleportTo() {
```
This method name is long and confusing. The `RequestTo` part is redundant since the whole event is regarding the event.

DrewAMitchell replied:
Absolutely. Will be fixed in the next commit.

mdcfe commented on Essentials/src/net/ess3/api/events/TPARequestEvent.java:
```diff
@@ -0,0 +1,55 @@
+		return !requestToTPAHere;
+	}
+	
+	public boolean isRequestToTeleportHere() {
```
We don't need both of these methods either - I'd suggest renaming `isRequestToTeleportHere` to `isTeleportHere` and removing `isRequestToTeleportTo`.

DrewAMitchell replied:
Taken into account!

mdcfe commented on Essentials/src/net/ess3/api/events/TPARequestEvent.java:
```diff
@@ -0,0 +1,55 @@
+	public TPARequestEvent(CommandSource requester, IUser requested, boolean tpaHere) {
+		super();
+		this.requester = requester;
+		this.requested = requested;
```
I'd suggest renaming this to `target` as it seems more in line with other references to the person receiving the request.

DrewAMitchell replied:
Done!

DrewAMitchell commented:
@drtshock I didn't use one of the current base events because I figured that the getRequester could be checked if it was an instanceof IUser, as the console command sender can trigger this event as well, so I made the requester a CommandSource.

khobbits commented:
I think I was just being lazy when I wrote StateChangeEvent, the reason was that there could be a dozen different triggers for commands, once you start including mods, console, rpc.  Easier to just work on the idea that if it was a player, we'd have a valid IUser, if not, null was as useful as anything else.

DrewAMitchell commented:
@md678685 @drtshock @SupaHam Re-review please!

DrewAMitchell commented:
My recent commit is still awaiting review; I will absolutely remove any TODO comments as requested. They're just there to allow me to remember what I wanted reviewed most from you guys. XD

mdcfe commented on Essentials/src/messages_fr.properties:
```diff
@@ -602,4 +603,4 @@ worthSet=Valeur cr\u00e9\u00e9e.
 years=ann\u00e9es
 youAreHealed=\u00a77Vous avez \u00e9t\u00e9 soign\u00e9.
-youHaveNewMail=\u00a7cVous avez {0} message(s) \! \u00a7fEntrez \u00a77/mail read\u00a7f pour voir votre courrier.
+youHaveNewMail=\u00a7cVous avez {0} message(s) \! \u00a7fEntrez \u00a77/mail read\u00a7f pour voir votre courrier.
```
Could you readd the new line at the end of this file?

mdcfe commented on Essentials/src/messages_it.properties:
```diff
@@ -602,4 +603,4 @@ worthSet=Valore definito
 years=anni
 youAreHealed=\u00a77Sei stato curato.
-youHaveNewMail=\u00a7cHai {0} messaggi\!\u00a7f digita \u00a77/mail read\u00a7f per consultare la tua mail.
+youHaveNewMail=\u00a7cHai {0} messaggi\!\u00a7f digita \u00a77/mail read\u00a7f per consultare la tua mail.
```
Same with newline here.

mdcfe commented on Essentials/src/messages_pl.properties:
```diff
@@ -602,4 +603,4 @@ worthSet=\u00a77Cena przedmiotu ustawiona.
 years=lat
 youAreHealed=\u00a77Zosta\u0142e\u015B uleczony.
-youHaveNewMail=\u00a77Masz\u00a7c {0} \u00a77wiadomosci\! Wpisz \u00a7c/mail read\u00a77 aby je przeczytac.
+youHaveNewMail=\u00a77Masz\u00a7c {0} \u00a77wiadomosci\! Wpisz \u00a7c/mail read\u00a77 aby je przeczytac.
```
Again, newline here.

mdcfe commented on Essentials/src/messages_zh.properties:
```diff
@@ -602,4 +603,4 @@ worthSet=\u00a76\u4ef7\u683c\u5df2\u8bbe\u7f6e
 years=\u5e74
 youAreHealed=\u00a76\u4f60\u5df2\u88ab\u6cbb\u7597
-youHaveNewMail=\u00a76\u4f60\u62e5\u6709 \u00a7c{0}\u00a76 \u6761\u6d88\u606f\uff01\u00a7r\u8f93\u5165 \u00a7c/mail read\u00a76 \u6765\u67e5\u770b
+youHaveNewMail=\u00a76\u4f60\u62e5\u6709 \u00a7c{0}\u00a76 \u6761\u6d88\u606f\uff01\u00a7r\u8f93\u5165 \u00a7c/mail read\u00a76 \u6765\u67e5\u770b
```
Last missing newline in a language file.

mdcfe commented on Essentials/src/messages_cs.properties:
```diff
@@ -602,4 +603,4 @@ worthSet=\u00a76Cena nastavena
 years=roky
 youAreHealed=\u00a76Byl/a jsi uzdraven/a.
-youHaveNewMail=\u00a76Mas \u00a7c{0} \u00a76zprav\! Napis \u00a7c/mail read\u00a76 pro precteni.
+youHaveNewMail=\u00a76Mas \u00a7c{0} \u00a76zprav\! Napis \u00a7c/mail read\u00a76 pro precteni.
```
I missed another missing newline here.

DrewAMitchell replied:
So, to clarify, you want an empty line at the end of the lang files?

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java:
```diff
@@ -1,13 +1,16 @@
-import static com.earth2me.essentials.I18n.tl;
+import org.bukkit.Server;
+
+import com.earth2me.essentials.User;
```
Any particular reason to reorder the imports, or was this just an IDE thing?

DrewAMitchell replied:
IDE thing, I'll fix it if you'd like me to - might fix it just because it's unnecessary.  :)

mdcfe commented on Essentials/src/net/ess3/api/events/TPARequestEvent.java:
```diff
@@ -0,0 +1,56 @@
+import net.ess3.api.IUser;
+
+
+public class TPARequestEvent extends Event implements Cancellable {
```
@drtshock @SupaHam Thoughts on having a `showCancellationMessage` boolean defaulting to `true` that determines whether EssentialsX shows its own "request cancelled" message, in case a plugin wants to display its own message instead?

mdcfe reviewed (changes requested):
There's a couple of really minor things like certain files being accidentally altered (presumably by the IDE), and I'm awaiting feedback from the others about whether to allow plugins to disable the cancellation message, but otherwise, this looks good.

DrewAMitchell commented:
I will be awaiting final thoughts before I make changes to the commit in the way of the new lines.
For clarification,
```yaml
someLanguageString="Test String; this is an example, so it doesn't matter."
#empty line
```
That empty line should be at the bottom of all the language files? Or is it just a cr/lf issue?

mdcfe commented:
I'm pretty sure the newlines are supposed to be at the bottom of the file, but you can leave them as-is for now.

Feel free to revert the changes in the order of imports back to how they were before - it would minimise the diff.

Still awaiting feedback from the others on my suggestion though.

Commit:
Attempts to implement #1682 - TPARequestEvent
Awaiting further feedback. I've implemented the event, and elected to send a message to the requester of the TPA when the event gets cancelled.

DrewAMitchell commented:
@md678685 You'll be happy to know that the latest squashed commit keeps your precious new-lines and minimizes the diff with the imports. I am now solely waiting on feedback from your companions.
@drtshock @SupaHam

mdcfe commented:
@drtshock @SupaHam Should we allow other plugins to manually disable our "request cancelled" message in case they want to show their own message?

SupaHam commented:
I don't think so. It'll be fine to have it print the official message and
the plugin to print theirs if necessary.

On Jan 25, 2018 8:10 PM, "md678685" <notifications@github.com> wrote:

> @drtshock <https://github.com/drtshock> @SupaHam
> <https://github.com/supaham> Should we allow other plugins to manually
> disable our "request cancelled" message in case they want to show their own
> message?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/EssentialsX/Essentials/pull/1727#issuecomment-360585675>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPV0UrQ9ulFFa26JTfOJX5IQc-oyK3iks5tON-ggaJpZM4RMbFM>
> .
>

DrewAMitchell commented:
@md678685 I'll remove my TODO's and comments and all that.

drtshock committed:
Merge branch '2.x' into TPARequest

mdcfe committed:
Merge branch '2.x' into TPARequest

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandtpaall.java:
```diff
@@ -30,7 +31,7 @@ public void run(final Server server, final CommandSource sender, final String co
 
     private void teleportAAllPlayers(final Server server, final CommandSource sender, final User target) {
-        sender.sendMessage(tl("teleportAAll"));
+        sender.sendMessage(tl("teleportAAll")); // TODO Take a look at whether we want to send this message before the loop, due to the TPARequestEvent implementation
```
Is this TODO still applicable? I think the initial message should be sent regardless of whether some are cancelled.

mdcfe committed:
Fix formatting

mdcfe reviewed (commented):
This is pretty much ready to merge, pending a decision on the TODO.

Ichbinjoe committed:
Merge branch '2.x' into TPARequest

Ichbinjoe committed:
Fix up l18s errors, remove todos

Ichbinjoe commented:
Looking for someone else to sign off on this (because I made changes)  then I'll merge this in.

mdcfe commented on Essentials/src/messages_cs.properties:
```diff
@@ -512,6 +512,7 @@ teleportInvalidLocation=Souradnice nemohou prekrocit 30 000 000
 teleportNewPlayerError=§4Teleportace noveho hrace selhala\!
 teleportRequest=§c{0}§c se chce teleportovat k tobe.
 teleportRequestAllCancelled=§6Vsechny nevyrizene zadosti o teleport byli zruseny.
+teleportRequestCancelled=§6Your teleport request to §c{0}§6 was cancelled.
```
Since the Crowdin integration, we now need to leave the translation files alone and only put new messages in `messages.properties`, else we'll get merge conflicts.

Edit: I'll fix this when I get time, but if anyone wants to fix this in the meantime feel free.

drtshock commented on Essentials/src/com/earth2me/essentials/commands/Commandtpa.java:
```diff
@@ -49,6 +50,11 @@ public void run(Server server, User user, String commandLabel, String[] args) th
         if (!player.isIgnoredPlayer(user)) {
+            TPARequestEvent tpaEvent = new TPARequestEvent(user.getSource(), player, false);
+            ess.getServer().getPluginManager().callEvent(tpaEvent);
+            if(tpaEvent.isCancelled()) {
```
formatting

drtshock commented on Essentials/src/com/earth2me/essentials/commands/Commandtpa.java:
```diff
@@ -49,6 +50,11 @@ public void run(Server server, User user, String commandLabel, String[] args) th
+            TPARequestEvent tpaEvent = new TPARequestEvent(user.getSource(), player, false);
+            ess.getServer().getPluginManager().callEvent(tpaEvent);
+            if(tpaEvent.isCancelled()) {
+                throw new Exception(tl("teleportRequestCancelled", player.getDisplayName())); // TODO Get approval on cancel message
```
Remove TODO before merge

drtshock commented on Essentials/src/com/earth2me/essentials/commands/Commandtpahere.java:
```diff
@@ -32,10 +33,15 @@ public void run(final Server server, final User user, final String commandLabel,
+            TPARequestEvent tpaEvent = new TPARequestEvent(user.getSource(), player, true);
+            ess.getServer().getPluginManager().callEvent(tpaEvent);
+            if (tpaEvent.isCancelled()) {
+                throw new Exception(tl("teleportRequestCancelled", player.getDisplayName()));
```
Not sure it should be our job to send a cancel message if another plugin is the one doing the cancelling, and likely has their own reasons for cancelling it.

mdcfe replied:
@drtshock I think other plugins can inform players if they want, but we should definitely inform the player ourselves because this is a condition where the teleportation has failed.

mdcfe committed:
Clean up for merge

mdcfe committed:
Merge remote-tracking branch 'origin/2.x' into TPARequest
